[ENH] introduce revised version of ETS

TonyBagnall opened the pull request:
part of #2833 

this replaces the current ETS with a numba version that is refactored. It is tested for equivalence, same tests pass and speed up over statsmodels is significant, cannot sensibly compare to statsforecast because that is tuning only, so obvs does more work, can compare that to autoETS

The first version was for some reason predicting an array rather than a float. Its still not clear to me how we use this with fit and predict, we might need a wrapper for an example. Basically fitting the model is so amazingly quick, refitting for each forecast is I believe worthwhile for interface compliance.

TonyBagnall commented:
output of the numba identical and runtime timing for main and this version

```python
def numba_compare():
    n =100
    import time
    import aeon.forecasting._ets as ets
    import aeon.forecasting._numba_ets as etsfast
    arr = np.random.rand(n)
    slow_ets = ets.ETSForecaster()
    fast_ets = etsfast.ETSForecaster()
    slow_ets.fit(arr)
    fast_ets.fit(arr)
    a1 = slow_ets.predict()
    a2 = fast_ets.predict()
    print(f" Slow = {a1} fast = {a2}")
    for n in range(100000,10000000,100000):
        arr = np.random.rand(n)

        ets_slow_time = time.time()
        slow_ets.fit(arr)
        ets_slow_time = time.time()-ets_slow_time
        ets_fast_time = time.time()
        fast_ets.fit(arr)
        ets_fast_time = time.time()-ets_fast_time
        ets_slow_pred = time.time()
        a1=slow_ets.predict()
        ets_slow_pred = time.time()-ets_slow_pred
        ets_fast_pred = time.time()
        a2=fast_ets.predict()
        ets_fast_pred = time.time()-ets_fast_pred
        equal = np.isclose(a1,a2, 4)
        print(f"{n},{ets_slow_time}, {ets_fast_time},{ets_slow_pred},,{ets_fast_pred}"
              f",{equal}")  # noqa
```

TonyBagnall commented:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/Tony/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/Tony/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"Aptos Narrow", sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl63
	{mso-number-format:Fixed;}
-->
</style>
</head>

<body link="#467886" vlink="#96607D">


2000000 | 4.12 | 0.04 | [ True]
-- | -- | -- | --
2100000 | 4.16 | 0.03 | [ True]
2200000 | 4.23 | 0.03 | [ True]
2300000 | 4.81 | 0.04 | [ True]
2400000 | 5.11 | 0.04 | [ True]
2500000 | 5.34 | 0.04 | [ True]
2600000 | 5.27 | 0.04 | [ True]
2700000 | 5.47 | 0.04 | [ True]
2800000 | 5.86 | 0.05 | [ True]
2900000 | 5.67 | 0.04 | [ True]
3000000 | 6.43 | 0.04 | [ True]
3100000 | 6.16 | 0.04 | [ True]
3200000 | 6.08 | 0.05 | [ True]
3300000 | 6.98 | 0.09 | [ True]
3400000 | 8.88 | 0.07 | [ True]
3500000 | 7.51 | 0.05 | [ True]
3600000 | 7.28 | 0.05 | [ True]
3700000 | 7.09 | 0.06 | [ True]
3800000 | 7.44 | 0.08 | [ True]
3900000 | 7.86 | 0.06 | [ True]
4000000 | 8.47 | 0.08 | [ True]
4100000 | 8.05 | 0.06 | [ True]
4200000 | 7.95 | 0.06 | [ True]
4300000 | 8.19 | 0.06 | [ True]
4400000 | 8.40 | 0.06 | [ True]
4500000 | 8.75 | 0.19 | [ True]



</body>

</html>

TonyBagnall commented:
much faster than statsmodels version, as equivalent as we can make it

```python

def statsmodels_compare(setup_func, random_seed, catch_errors):
    """Run both our statsforecast and our implementation and crosschecks."""
    import warnings
    warnings.filterwarnings("ignore")
    random.seed(random_seed)
    (
        y,
        m,
        error,
        trendtype,
        seasontype,
        alpha,
        beta,
        gamma,
        phi,
    ) = setup_func()
    # tsml-eval implementation
    start = time.perf_counter()
    f1 = etsfast.ETSForecaster(
        error,
        trendtype,
        seasontype,
        m,
        alpha,
        beta,
        gamma,
        phi,
        1,
    )
    for n in range(10000,1000000,10000):
        y = np.random.rand(n)
        start = time.perf_counter()
        f1 = etsfast.ETSForecaster(
            error,
            trendtype,
            seasontype,
            m,
            alpha,
            beta,
            gamma,
            phi,
            1,
        )
        f1.fit(y)
        aeon_time = time.perf_counter()-start
        from statsmodels.tsa.holtwinters import ExponentialSmoothing
        start = time.perf_counter()
        ets_model = ExponentialSmoothing(
            y[m:],
            trend="add" if trendtype == 1 else "mul" if trendtype == 2 else None,
            damped_trend=(phi != 1 and trendtype != 0),
            seasonal="add" if seasontype == 1 else "mul" if seasontype == 2 else None,
            seasonal_periods=m if seasontype != 0 else None,
            initialization_method="known",
            initial_level=f1.level_,
            initial_trend=f1.trend_ if trendtype != 0 else None,
            initial_seasonal=f1.seasonality_ if seasontype != 0 else None,
        )
        results = ets_model.fit(
            smoothing_level=alpha,
            smoothing_trend=(
                beta / alpha if trendtype != 0 else None
            ),  # statsmodels uses beta*=beta/alpha
            smoothing_seasonal=gamma if seasontype != 0 else None,
            damping_trend=phi if trendtype != 0 else None,
            optimized=False,
        )
        sm_time = time.perf_counter()-start
        print(f"{n},{aeon_time}, {sm_time:0.20f}")  # noqa
 ```

TonyBagnall commented:
length, aeon, statsmodels
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/Tony/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/Tony/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"Aptos Narrow", sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
-->
</style>
</head>

<body link="#467886" vlink="#96607D">


700000 | 0.011834 | 3.057566
-- | -- | --
710000 | 0.012088 | 3.139149
720000 | 0.013966 | 3.217208
730000 | 0.012504 | 3.63404
740000 | 0.015816 | 3.24126
750000 | 0.013041 | 3.542698
760000 | 0.013758 | 3.303388
770000 | 0.013628 | 3.537901
780000 | 0.013103 | 3.474043
790000 | 0.013888 | 3.391304
800000 | 0.01361 | 3.626867
810000 | 0.013806 | 3.450841
820000 | 0.013852 | 3.513971
830000 | 0.017863 | 3.660681
840000 | 0.014946 | 3.756113
850000 | 0.015857 | 3.80876
860000 | 0.022167 | 3.932306



</body>

</html>

TonyBagnall commented:
it cannot be compared to the statsforecast model, because you cannot set alpha, beta, you can only auto tune them

MatthewMiddlehurst commented on aeon/forecasting/_ets.py:
```diff
@@ -25,86 +29,84 @@
-        gamma: float = 0.01,
-        phi: float = 0.99,
-        horizon: int = 1,
+        error_type=ADDITIVE,
+        trend_type=NONE,
+        seasonality_type=NONE,
+        seasonal_period=1,
+        alpha=0.1,
+        beta=0.01,
+        gamma=0.01,
+        phi=0.99,
+        horizon=1,
```
This and the docstring are out of order.

Also I see why these are ints now, but still think str would be better. Don't think enums fit well IMO. Probably more an issue with the language

MatthewMiddlehurst commented on aeon/forecasting/_ets.py:
```diff
@@ -25,86 +29,84 @@
         self.phi = phi
-        self.mean_sq_err_ = 0
-        self.likelihood_ = 0
+        self.forecast_val_ = 0.0
+        self.level_ = 0.0
+        self.trend_ = 0.0
+        self.seasonality_ = None
+        self._beta = beta
+        self._gamma = gamma
+        self.error_type = error_type
+        self.trend_type = trend_type
+        self.seasonality_type = seasonality_type
+        self.seasonal_period = seasonal_period
+        self._seasonal_period = seasonal_period
+        self.n_timepoints_ = 0
+        self.avg_mean_sq_err_ = 0
+        self.liklihood_ = 0
+        self.k_ = 0
+        self.aic_ = 0
```
Public attributes should ideally be documented in the docstring. 

Setting all these is also rather disorganised.

MatthewMiddlehurst commented on aeon/forecasting/_ets.py:
```diff
@@ -25,86 +29,84 @@
 
     def __init__(
         self,
-        error_type: int = ADDITIVE,
-        trend_type: int = NONE,
-        seasonality_type: int = NONE,
-        seasonal_period: int = 1,
-        alpha: float = 0.1,
-        beta: float = 0.01,
-        gamma: float = 0.01,
-        phi: float = 0.99,
-        horizon: int = 1,
```
Why are we removing these?

alexbanwell1 replied:
Mostly because this was a commit someone else added at some point - they never made it into my code because at some point the ets version in main was producing incorrect results, and I had to fix it! I've added them back in now!

MatthewMiddlehurst commented on aeon/forecasting/_ets.py:
```diff
@@ -216,39 +242,59 @@ def _fit_numba(
+    )
+
+
+@njit(nogil=NOGIL, cache=CACHE)
```
What are these vars? we dont use them anywhere else.

TonyBagnall replied:
I'll sere what happens if I change them to our defaults

TonyBagnall replied:
it is no longer based on statsforecast as I understand it its been completely refactored @alexbanwell1 thloughts?

MatthewMiddlehurst reviewed (commented):
The current version appears to use numba already? Not quite sure what the code changes are. It does not appear to be just numbafying what is current there at least.

alexbanwell1 commented on aeon/forecasting/_ets.py:
```diff
@@ -25,86 +29,84 @@
+    trend_type : int
+        The type of trend model; one of None(0), additive(1) or multiplicative(2).
+    seasonality_type : int
+        The type of seasonality model; one of None(0), additive(1) or multiplicative(2).
```
I've updated it to be strings. Will mean AutoETS will need an update when we review that. I've left the constants, as its so much easier to code with and decreases bugs!

TonyBagnall commented:
I have changed it so that you can input either strings or ints for error, trend and seasonality, these are now validated in fit and converted to ints for numba efficiency. Added a test of this conversion

hadifawaz1999 commented on aeon/forecasting/_ets.py:
```diff
@@ -1,110 +1,131 @@
-        Either NONE (0), ADDITIVE (1) or MULTIPLICATIVE (2).
-    seasonality_type : int, default = 0
-        Either NONE (0), ADDITIVE (1) or MULTIPLICATIVE (2).
+    error_type : string or in default='additive'
```
typo "int" not "in"

hadifawaz1999 reviewed (changes requested):
Looks great ! just some minor docs stuff from my side

hadifawaz1999 reviewed (approved):
lgtm
